File: 44-Pentagon_Numbers/main.py
def is_pentagonal(x):
    """Check if a number x is a pentagonal number"""
    n = (1 + (1 + 24 * x) ** 0.5) / 6
    return n == int(n)

def pentagonal(n):
    """Generate the n-th pentagonal number"""
    return n * (3 * n - 1) // 2

def find_min_pentagonal_pair():
    pentagonals = []
    n = 1
    # We brute force the answer
    while True:
        P_n = pentagonal(n)
        pentagonals.append(P_n)
        for P_j in pentagonals[:-1]:
            P_k = P_n
            if is_pentagonal(P_k - P_j) and is_pentagonal(P_k + P_j):  # The first one to be find will be the minimized one since the difference will only grow while we iterate
                return P_k - P_j
        n += 1

D = find_min_pentagonal_pair()

print("The answer is: " + str(D))

File: 44-Pentagon_Numbers/main.cpp
#include <iostream>
#include <cmath>
#include <vector>

bool isPentagonal(long x) {
    double n = (1 + sqrt(1 + 24 * x)) / 6;
    return n == (long) n;
}

long pentagonal(int n) {
    return n * (3LL * n - 1) / 2;
}

int main() {
    std::vector<long> pentagonals;
    int n = 1;

    while (true) {
        long P_n = pentagonal(n);
        pentagonals.push_back(P_n);

        for (auto P_j : pentagonals) {
            if (P_j == P_n) continue;
            long P_k = P_n;
            if (isPentagonal(P_k - P_j) && isPentagonal(P_k + P_j)) {
                std::cout << "D = " << P_k - P_j << std::endl;
                return 0;
            }
        }

        n++;
    }

    return 0;
}
